aws/rust-runtime/aws-inlineable/src/body_with_checksum.rs

```diff
@@ -0,0 +1,195 @@
+    );
+
+    Ok(())
+}
+
+/// Errors related to constructing checksum-validated HTTP requests
+#[derive(Debug)]
+pub enum ChecksumValidatedRequestError {
+    /// Only request bodies with a known size can be checksum validated
+    UnsizedRequestBody,
+}
+
+impl std::fmt::Display for ChecksumValidatedRequestError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::UnsizedRequestBody => write!(
+                f,
+                "Only request bodies with a known size can be checksum validated."
+            ),
+        }
+    }
+}
+
+impl std::error::Error for ChecksumValidatedRequestError {}
+
+/// Given an `http::request::Builder`, `SdkBody`, and a checksum algorithm name, return a
+/// `Request<SdkBody>` with checksum trailers where the content is `aws-chunked` encoded.
+pub fn build_checksum_validated_request_with_streaming_body(
+    request: &mut http::request::Request<aws_smithy_http::body::SdkBody>,
+    checksum_algorithm: &str,
+) -> Result<(), aws_smithy_http::operation::BuildError> {
+    use http_body::Body;
+
+    let original_body_size = request.body().size_hint().exact().ok_or_else(|| {
+        aws_smithy_http::operation::BuildError::Other(Box::new(
+            ChecksumValidatedRequestError::UnsizedRequestBody,
+        ))
+    })?;
+    let body = std::mem::replace(request.body_mut(), aws_smithy_http::body::SdkBody::taken());
+    let body = aws_smithy_checksums::body::ChecksumBody::new(body, checksum_algorithm);
+    let checksum_trailer_name = body.trailer_name();
+    let aws_chunked_body_options = aws_http::content_encoding::AwsChunkedBodyOptions::new()
+        .with_stream_length(original_body_size)
+        .with_trailer_len(body.trailer_length());
+
+    let body = aws_http::content_encoding::AwsChunkedBody::new(body, aws_chunked_body_options);
+    let encoded_content_length = body.size_hint().exact().ok_or_else(|| {
+        aws_smithy_http::operation::BuildError::Other(Box::new(
+            ChecksumValidatedRequestError::UnsizedRequestBody,
+        ))
+    })?;
+    let headers = request.headers_mut();
+    headers.insert(
+        http::header::CONTENT_LENGTH,
+        http::HeaderValue::from(encoded_content_length),
+    );
+    headers.insert(
+        http::header::HeaderName::from_static("x-amz-decoded-content-length"),
+        http::HeaderValue::from(original_body_size),
+    );
+    headers.insert(
+        http::header::HeaderName::from_static("x-amz-trailer"),
+        checksum_trailer_name.into(),
+    );
+    headers.insert(
+        http::header::CONTENT_ENCODING,
+        http::HeaderValue::from_str(aws_http::content_encoding::header_value::AWS_CHUNKED)
+            .map_err(|err| aws_smithy_http::operation::BuildError::Other(Box::new(err)))?,
+    );
+
+    let mut body =
+        aws_smithy_http::body::SdkBody::from_dyn(http_body::combinators::BoxBody::new(body));
```

Comment:
this will make these requests all non-retryable which probably makes this unmergeable as-is (can review the rest of the PR, though)

Reply:
I fixed this.